# Part 1: Data loading and preprocessing

Первые шаги включают загрузку данных и их предварительный анализ.

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
data = pd.read_csv('intern_task.csv')
data.head()

rank  query_id  feature_0  feature_1  feature_2  feature_3  feature_4  \
0     0        10        1.0        0.0        1.0        3.0        3.0   
1     1        10        3.0        0.0        3.0        0.0        3.0   
2     0        10        3.0        0.0        2.0        0.0        3.0   
3     1        10        3.0        0.0        3.0        0.0        3.0   
4     2        10        3.0        0.0        3.0        1.0        3.0   

   feature_5  feature_6  feature_7  ...  feature_134  feature_135  \
0   0.333333        0.0   0.333333  ...          0.0     0.000000   
1   1.000000        0.0   1.000000  ...          0.0     0.000000   
2   1.000000        0.0   0.666667  ...          0.0     0.000000   
3   1.000000        0.0   1.000000  ...          0.0     0.000000   
4   1.000000        0.0   1.000000  ...        273.0    79.670665   

   feature_136  feature_137  feature_138  feature_139  feature_140  \
0     0.454545     0.890238     8.655534     1.000000     0.077778   
1     0.000000     0.773976    23.130514     0.000000     0.027826   
2     0.000000     0.918308    13.351339     0.000000     0.014925   
3     0.000000     0.975355    18.240926     0.000000     0.053140   
4     0.200000     0.990119    31.786048     0.333333     0.046512   

   feature_141  feature_142  feature_143  
0     0.002222          1.0     0.333333  
1     0.000430         44.0    14.666667  
2     0.000104         22.0     7.333333  
3     0.000255          8.0     2.666667  
4     0.000307         24.0     8.000000  

[5 rows x 146 columns]

In [3]:
import lightgbm as lgb
print(lgb.__version__)

4.3.0


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235258 entries, 0 to 235257
Columns: 146 entries, rank to feature_143
dtypes: float64(140), int64(6)
memory usage: 262.1 MB


In [5]:
data.describe()

rank       query_id      feature_0      feature_1  \
count  235258.000000  235258.000000  235258.000000  235258.000000   
mean        0.677869   14828.413401       1.911960       0.206233   
std         0.830033    8193.945170       1.237374       0.579089   
min         0.000000      10.000000       0.000000       0.000000   
25%         0.000000    8215.000000       1.000000       0.000000   
50%         0.000000   14935.000000       2.000000       0.000000   
75%         1.000000   21580.000000       3.000000       0.000000   
max         4.000000   29995.000000      31.000000      18.000000   

           feature_2      feature_3      feature_4      feature_5  \
count  235258.000000  235258.000000  235258.000000  235258.000000   
mean        1.189847       0.550272       1.960082       0.803964   
std         1.037233       0.790947       1.203534       0.339955   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       1.000000       0.666667   
50%         1.000000       0.000000       2.000000       1.000000   
75%         2.000000       1.000000       3.000000       1.000000   
max        27.000000       9.000000      31.000000       1.000000   

           feature_6      feature_7  ...   feature_134   feature_135  \
count  235258.000000  235258.000000  ...  2.352580e+05  2.352580e+05   
mean        0.097557       0.541174  ...  4.763432e+02  1.046698e+04   
std         0.266035       0.419973  ...  1.954840e+04  2.649584e+06   
min         0.000000       0.000000  ...  0.000000e+00  0.000000e+00   
25%         0.000000       0.000000  ...  0.000000e+00  0.000000e+00   
50%         0.000000       0.500000  ...  0.000000e+00  0.000000e+00   
75%         0.000000       1.000000  ...  1.000000e+00  4.000000e-01   
max         1.000000       1.000000  ...  1.731553e+06  8.000000e+08   

         feature_136    feature_137    feature_138    feature_139  \
count  235258.000000  235258.000000  235258.000000  235258.000000   
mean        0.070299       0.740710      11.067947       0.281747   
std         0.100721       0.352800      15.336861       0.392089   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.620522       0.000000       0.000000   
50%         0.000000       0.926779       9.448900       0.000000   
75%         0.142857       0.999613      16.146733       0.500000   
max         0.818182       1.000000    2506.799764       1.000000   

         feature_140    feature_141    feature_142    feature_143  
count  235258.000000  235258.000000  235258.000000  235258.000000  
mean        0.027033       0.000269      22.457910       9.560379  
std         0.033351       0.002088      63.708018      14.347378  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.006703       0.000000       4.000000       2.000000  
50%         0.017761       0.000008      12.000000       5.500000  
75%         0.034954       0.000109      28.000000      12.000000  
max         0.711261       0.250000   15074.000000    1552.000000  

[8 rows x 146 columns]

In [6]:
print(data.head())

   rank  query_id  feature_0  feature_1  feature_2  feature_3  feature_4  \
0     0        10        1.0        0.0        1.0        3.0        3.0   
1     1        10        3.0        0.0        3.0        0.0        3.0   
2     0        10        3.0        0.0        2.0        0.0        3.0   
3     1        10        3.0        0.0        3.0        0.0        3.0   
4     2        10        3.0        0.0        3.0        1.0        3.0   

   feature_5  feature_6  feature_7  ...  feature_134  feature_135  \
0   0.333333        0.0   0.333333  ...          0.0     0.000000   
1   1.000000        0.0   1.000000  ...          0.0     0.000000   
2   1.000000        0.0   0.666667  ...          0.0     0.000000   
3   1.000000        0.0   1.000000  ...          0.0     0.000000   
4   1.000000        0.0   1.000000  ...        273.0    79.670665   

   feature_136  feature_137  feature_138  feature_139  feature_140  \
0     0.454545     0.890238     8.655534     1.000000     

In [7]:
print("Missing values in each column:\n", data.isnull().sum())

Missing values in each column:
 rank           0
query_id       0
feature_0      0
feature_1      0
feature_2      0
              ..
feature_139    0
feature_140    0
feature_141    0
feature_142    0
feature_143    0
Length: 146, dtype: int64


Данные необходимо подготовить для обучения модели. Это включает в себя scaling признаков и разделение на обучающую и тестовую выборки.

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
feature_columns = [col for col in data.columns if col.startswith('feature_')]
data[feature_columns] = scaler.fit_transform(data[feature_columns])

In [9]:
data.head()

rank  query_id  feature_0  feature_1  feature_2  feature_3  feature_4  \
0     0        10  -0.737015  -0.356135  -0.183032   3.097217   0.864056   
1     1        10   0.879315  -0.356135   1.745178  -0.695715   0.864056   
2     0        10   0.879315  -0.356135   0.781073  -0.695715   0.864056   
3     1        10   0.879315  -0.356135   1.745178  -0.695715   0.864056   
4     2        10   0.879315  -0.356135   1.745178   0.568595   0.864056   

   feature_5  feature_6  feature_7  ...  feature_134  feature_135  \
0  -1.384395   -0.36671  -0.494892  ...    -0.024367     -0.00395   
1   0.576653   -0.36671   1.092516  ...    -0.024367     -0.00395   
2   0.576653   -0.36671   0.298813  ...    -0.024367     -0.00395   
3   0.576653   -0.36671   1.092516  ...    -0.024367     -0.00395   
4   0.576653   -0.36671   1.092516  ...    -0.010402     -0.00392   

   feature_136  feature_137  feature_138  feature_139  feature_140  \
0     3.814976     0.423834    -0.157295     1.831865     1.521570   
1    -0.697956     0.094293     0.786510    -0.718582     0.023786   
2    -0.697956     0.503398     0.148883    -0.718582    -0.363043   
3    -0.697956     0.665096     0.467696    -0.718582     0.782813   
4     1.287736     0.706945     1.350873     0.131566     0.584076   

   feature_141  feature_142  feature_143  
0     0.935567    -0.336817    -0.643119  
1     0.077291     0.338139     0.355905  
2    -0.078846    -0.007188    -0.155224  
3    -0.006525    -0.226941    -0.480487  
4     0.018381     0.024206    -0.108757  

[5 rows x 146 columns]

In [10]:
query_ids = data['query_id'].unique()
train_ids, test_ids = train_test_split(query_ids, test_size=0.2, random_state=42)

In [11]:
train_data = data[data['query_id'].isin(train_ids)]
test_data = data[data['query_id'].isin(test_ids)]

print("Train data shape:", train_data.shape)
print("Test data shape:", test_data.shape)

Train data shape: (188989, 146)
Test data shape: (46269, 146)


# Part 2: Training LambdaMART

In [12]:
import lightgbm as lgb
from sklearn.metrics import ndcg_score
import numpy as np

# Prepare the datasets for LightGBM
train_group = train_data.groupby('query_id').size().to_numpy()
test_group = test_data.groupby('query_id').size().to_numpy()

# Features and labels
X_train = train_data[feature_columns]
y_train = train_data['rank']
X_test = test_data[feature_columns]
y_test = test_data['rank']

In [13]:
ranker = lgb.LGBMRanker(
    objective='lambdarank',
    metric='ndcg',
    learning_rate=0.1,
    n_estimators=100,  
    num_leaves=31,    
    seed=42,
    verbose=10,
    early_stopping_rounds=50
)

In [14]:
ranker.fit(
    X_train, y_train,
    group=train_group,
    eval_set=[(X_test, y_test)],
    eval_group=[test_group],
    eval_at=[5, 10, 20]
)

[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Dataset::GetMultiBinFromSparseFeatures: sparse rate 0.857021
[LightGBM] [Debug] Dataset::GetMultiBinFromAllFeatures: sparse rate 0.185768
[LightGBM] [Debug] init for col-wise cost 0.012053 seconds, init for row-wise cost 0.027542 seconds
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.025872 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Debug] Using Dense Multi-Val Bin
[LightGBM] [Info] Total Bins 26741
[LightGBM] [Info] Number of data points in the train set: 188989, number of used features: 140
[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Trained a tree with leaves = 31 and depth = 8
Training 

LGBMRanker(early_stopping_rounds=50, metric='ndcg', objective='lambdarank',
           seed=42, verbose=10)

In [15]:
test_pred = ranker.predict(X_test)

In [16]:
def calculate_ndcg(y_true, y_score, k=None):
    true_relevance = np.asarray([y_true])
    scores = np.asarray([y_score])
    if k is None:
        ndcg = ndcg_score(true_relevance, scores)
    else:
        ndcg = ndcg_score(true_relevance, scores, k=k)
    return ndcg

In [17]:
ndcg_at_5 = calculate_ndcg(y_test, test_pred, k=5)
print("NDCG@5:", ndcg_at_5)

NDCG@5: 0.9576049743407978


Наша LambdaMART модель с определенными параметрами выбила NDCG@5 = 0.958. 

# Part 3: Hyperparameter tuning for LambdaMART

Ищем лучшие гиперпараметры для модели, оптимизируя NDCG@5.

Стоит отметить, что так как мы используем metric='ndcg', то мы не можем воспользоваться GridSearchCV прямо из sklearn.model_selection.

Поэтому мы напишем свой кастомный подбор гиперпараметров.

In [18]:
learning_rates = [0.01, 0.05, 0.1]
n_estimators_list = [50, 100, 200]
num_leaves_list = [15, 31, 50]
min_child_samples_list = [10, 20, 30]

best_ndcg_score = 0
best_params = {}

for learning_rate in learning_rates:
    for n_estimators in n_estimators_list:
        for num_leaves in num_leaves_list:
            for min_child_samples in min_child_samples_list:
                ranker = lgb.LGBMRanker(objective='lambdarank', metric='ndcg', 
                                         learning_rate=learning_rate, 
                                         n_estimators=n_estimators, 
                                         num_leaves=num_leaves,
                                         min_child_samples=min_child_samples,
                                         seed=42,
                                         verbose=10,
                                         early_stopping_rounds=50)
                ranker.fit(X_train, y_train, group=train_group, eval_set=[(X_test, y_test)], 
                           eval_group=[test_group], eval_at=[5, 10, 20])
                
                test_pred = ranker.predict(X_test)
                ndcg_at_5 = calculate_ndcg(y_test, test_pred, k=5)
                ndcg_at_10 = calculate_ndcg(y_test, test_pred, k=10)
                ndcg_at_20 = calculate_ndcg(y_test, test_pred, k=20)
                
                if ndcg_at_5 > best_ndcg_score:
                    best_ndcg_score = ndcg_at_5
                    best_params = {'learning_rate': learning_rate,
                                   'n_estimators': n_estimators,
                                   'num_leaves': num_leaves,
                                   'min_child_samples': min_child_samples}
                
print("Best Parameters:", best_params)
print("Best NDCG@5 Score:", best_ndcg_score)

[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Dataset::GetMultiBinFromSparseFeatures: sparse rate 0.857021
[LightGBM] [Debug] Dataset::GetMultiBinFromAllFeatures: sparse rate 0.185768
[LightGBM] [Debug] init for col-wise cost 0.016085 seconds, init for row-wise cost 0.032407 seconds
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.029066 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Debug] Using Dense Multi-Val Bin
[LightGBM] [Info] Total Bins 26741
[LightGBM] [Info] Number of data points in the train set: 188989, number of used features: 140
[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Trained a tree with leaves = 15 and depth = 7
Training 

Как показал наш процесс подбора гиперпараметров, NDCG@5 максимальный при данном наборе гиперпараметров: 

Best Parameters: {'learning_rate': 0.05, 'n_estimators': 100, 'num_leaves': 15, 'min_child_samples': 10}

Best NDCG@5 Score: 0.9671987306219144

Теперь натренируем нашу модель с данными гиперпараметрами.

In [19]:
# Train the model with the best parameters
best_ranker = lgb.LGBMRanker(objective='lambdarank', metric='ndcg', 
                              learning_rate=0.05, n_estimators=100, 
                              num_leaves=15, min_child_samples=10,
                              seed=42, verbose=10, early_stopping_rounds=50)

best_ranker.fit(X_train, y_train, group=train_group, eval_set=[(X_test, y_test)], 
                eval_group=[test_group], eval_at=[5, 10, 20])

test_pred = best_ranker.predict(X_test)

[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Dataset::GetMultiBinFromSparseFeatures: sparse rate 0.857021
[LightGBM] [Debug] Dataset::GetMultiBinFromAllFeatures: sparse rate 0.185768
[LightGBM] [Debug] init for col-wise cost 0.011915 seconds, init for row-wise cost 0.031517 seconds
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.025333 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Debug] Using Dense Multi-Val Bin
[LightGBM] [Info] Total Bins 26741
[LightGBM] [Info] Number of data points in the train set: 188989, number of used features: 140
[LightGBM] [Warning] early_stopping_round is set=50, early_stopping_rounds=50 will be ignored. Current value: early_stopping_round=50
[LightGBM] [Debug] Trained a tree with leaves = 15 and depth = 7
Training 

# Part 4: Final model evaluation

In [20]:
# Calculate the metrics
ndcg_at_5 = calculate_ndcg(y_test, test_pred, k=5)
ndcg_at_10 = calculate_ndcg(y_test, test_pred, k=10)
ndcg_at_20 = calculate_ndcg(y_test, test_pred, k=20)
ndcg = calculate_ndcg(y_test, test_pred)

print("NDCG@5:", ndcg_at_5)
print("NDCG@10:", ndcg_at_10)
print("NDCG@20:", ndcg_at_20)
print("NDCG:", ndcg)

NDCG@5: 0.9671987306219144
NDCG@10: 0.8901528648019593
NDCG@20: 0.8375158147770515
NDCG: 0.9210906413898928


Модель показала хорошие результаты по всем метрикам. 